nifi-nar-bundles/nifi-cypher-bundle/nifi-neo4j-cypher-service/src/main/java/org/apache/nifi/cypher/Neo4JCypherClientService.java

```diff
@@ -0,0 +1,304 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class Neo4JCypherClientService extends AbstractControllerService implements CypherClientService {
+    public static final PropertyDescriptor CONNECTION_URL = new PropertyDescriptor.Builder()
+            .name("neo4j-connection-url")
+            .displayName("Neo4j Connection URL")
+            .description("Neo4J endpoing to connect to.")
+            .required(true)
+            .defaultValue("bolt://localhost:7687")
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor USERNAME = new PropertyDescriptor.Builder()
+            .name("neo4j-username")
+            .displayName("Username")
+            .description("Username for accessing Neo4J")
+            .required(true)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PASSWORD = new PropertyDescriptor.Builder()
+            .name("neo4j-password")
+            .displayName("Password")
+            .description("Password for Neo4J user. A dummy non-blank password is required even if it disabled on the server.")
+            .required(true)
+            .sensitive(true)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+    public static AllowableValue LOAD_BALANCING_STRATEGY_ROUND_ROBIN = new AllowableValue(Config.LoadBalancingStrategy.ROUND_ROBIN.name(), "Round Robin", "Round Robin Strategy");
+
+    public static AllowableValue LOAD_BALANCING_STRATEGY_LEAST_CONNECTED = new AllowableValue(Config.LoadBalancingStrategy.LEAST_CONNECTED.name(), "Least Connected", "Least Connected Strategy");
+
+    protected static final PropertyDescriptor LOAD_BALANCING_STRATEGY = new PropertyDescriptor.Builder()
+            .name("neo4j-load-balancing-strategy")
+            .displayName("Load Balancing Strategy")
+            .description("Load Balancing Strategy (Round Robin or Least Connected)")
+            .required(false)
+            .defaultValue(LOAD_BALANCING_STRATEGY_ROUND_ROBIN.getValue())
+            .allowableValues(LOAD_BALANCING_STRATEGY_ROUND_ROBIN, LOAD_BALANCING_STRATEGY_LEAST_CONNECTED)
+            .build();
+
+    public static final PropertyDescriptor CONNECTION_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("neo4j-max-connection-time-out")
+            .displayName("Neo4J Max Connection Time Out (seconds)")
+            .description("The maximum time for establishing connection to the Neo4j")
+            .defaultValue("5 seconds")
+            .required(true)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
```

Comment:
Maybe add support for EL with Variable Registry for the connection settings? Might need to update the `CypherClientService` to take an attribute map, see `DBCPService` for an example.

Reply:
CypherClientService is more like the HBase client service we have than a JDBC wrapper. OpenCypher is a thing wrapper on top of Gremlin, and there's almost nothing in common in terms of configuration between the two clients. So I'm very wary about going that route.

Reply:
I'll also add that this is a Neo4J vs JanusGraph issue too because w/ JanusGraph you typically back it with HBase and ElasticSearch to build a really big data set. You can have multiple graphs in the same cluster, but what's the point when you can theoretically scale as much as HBase and Elastic will allow (which should be substantially higher than Neo4J right now).